src/full/Agda/Syntax/Parser/LexActions.hs

```diff
@@ -191,9 +192,24 @@ symbol s = withInterval_ (TokSymbol s)
 -- | Parse a number.
 
 number :: String -> Integer
-number str = read $ case str of
-  '0' : 'x' : num -> str
-  _               -> concat $ wordsBy ('_' ==) str
+number str = case str of
+    '0' : 'x' : num -> parseNumber 16 num
+    '0' : 'b' : num -> parseNumber 2  num
+    num             -> parseNumber 10 num
+    where
+        parseNumber :: Integer -> String -> Integer
+        parseNumber radix = foldl' (addDigit radix)  0
+
+        addDigit :: Integer -> Integer -> Char -> Integer
+        addDigit radix n '_' = n
+        addDigit radix n c   = n * radix + fromIntegral (parseDigit c)
+
+        parseDigit :: Char -> Int
```

Comment:
May be worth adding a comment stating that `Agda.Syntax.Parser.Lexer` already
enforces that the characters are in the right ranges so there is no need to
double check that the user has not written e.g. `0b1DEAD`.

Reply:
Good idea, done